ostd/libs/linux-bzimage/setup/src/x86/amd64_efi/efi.rs

```diff
@@ -55,9 +56,12 @@ fn efi_phase_boot(
 
     uefi_services::println!("[EFI stub] Relocations applied.");
 
-    uefi_services::println!("[EFI stub] Loading payload.");
+    uefi_services::println!("[EFI stub] Decompressing payload.");
     let payload = unsafe { crate::get_payload(&*boot_params_ptr) };
-    crate::loader::load_elf(payload);
+    let kernel = decompress_payload(payload);
```

Comment:
You need to tell the type of the payload (by reading the magic number in the payload) before decompressing it.

https://www.kernel.org/doc/html/v5.6/x86/boot.html says:

> The payload may be compressed. The format of both the compressed and uncompressed data should be determined using the standard magic numbers. The currently supported compression formats are gzip (magic numbers 1F 8B or 1F 9E), bzip2 (magic number 42 5A), LZMA (magic number 5D 00), XZ (magic number FD 37), and LZ4 (magic number 02 21). The uncompressed payload is currently always ELF (magic number 7F 45 4C 46).